File: resources/corpus.ts
import * as fs from 'fs';
import * as path from 'path';

// Define valid characters as a string
const validChars = "abcdefghijklmnopqrstuvwxyz ".split('');

// Function to find intersection of two lists
function intersection(lst1: string[], lst2: string[]): string[] {
    const temp = new Set(lst2);
    return lst1.filter(value => temp.has(value));
}

// Function to find n-grams (bigrams, etc.) in a string
function findNgrams(str: string, n: number): [string, string][] {
    const inputList = intersection(str.toLowerCase().split(''), validChars);
    const result: [string, string][] = [];
    for (let i = 0; i < inputList.length - n + 1; i++) {
        result.push(inputList.slice(i, i + n) as [string, string]);
    }
    return result;
}

// Function to calculate unigrams
function unigram(language: string, file: string) {
    const text = fs.readFileSync(file, 'utf-8').toLowerCase();
    const occurrence = new Map<string, number>();
    const filteredText = intersection(text.split(''), validChars);

    filteredText.forEach(char => {
        occurrence.set(char, (occurrence.get(char) || 0) + 1);
    });

    const count = filteredText.length;
    const filename = `unigram_${language}.txt`;

    const dump = fs.createWriteStream(filename);
    validChars.forEach(char => {
        const prob = (occurrence.get(char) || 0) / count;
        dump.write(`'${char}': ${prob.toExponential(6)},\n`);
    });

    dump.end();
    console.log("unigram", language, file, count);
}

// Function to calculate bigrams
function bigram(language: string, file: string) {
    const text = fs.readFileSync(file, 'utf-8').toLowerCase();
    const occurrence = new Map<string, number>();
    const bigrams = findNgrams(text, 2);

    bigrams.forEach(([char1, char2]) => {
        const key = char1 + char2;
        occurrence.set(key, (occurrence.get(key) || 0) + 1);
    });

    const count = bigrams.length;
    const filename = `bigram_${language}.txt`;

    const dump = fs.createWriteStream(filename);
    validChars.forEach(char1 => {
        validChars.forEach(char2 => {
            const key = char1 + char2;
            const prob = (occurrence.get(key) || 0) / count;
            dump.write(`'${key}': ${prob.toExponential(6)},\n`);
        });
    });

    dump.end();
    console.log("bigram", language, file, count);
}

// Corpus of languages and corresponding files
const corpus: { [key: string]: string } = {
    "danish": "../data/dan_news_2008_30K-sentences.txt",
    "english": "../data/eng_news_2016_30K-sentences.txt",
    "french": "../data/fra_news_2010_30K-sentences.txt",
    "german": "../data/deu_news_2015_30K-sentences.txt",
    "swedish": "../data/swe_news_2007_30K-sentences.txt"
};

// Loop over each language and process bigrams (unigram can be called similarly)
for (const language in corpus) {
    unigram(language, corpus[language]);
    bigram(language, corpus[language]);
}

File: scripts/corpus.py
import collections
import math
from decimal import Decimal

valid_chars = list("abcdefghijklmnopqrstuvwxyz ")

def intersection(lst1, lst2): 
	temp = set(lst2) 
	lst3 = [value for value in lst1 if value in temp] 
	return lst3 

def find_ngrams(str, n):
	input_list = intersection(list(str.lower()), valid_chars)
	return zip(*[input_list[i:] for i in range(n)])

def unigram(text, file):
	text = open(file).read().lower()
	occurance = collections.Counter(intersection(list(text), valid_chars))
	count = len(intersection(list(text), valid_chars))

	filename = ("unigram_%s.txt" % language)
	dump = open(filename, "w")
	for char in valid_chars:
		dump.write("'%s': %e,\n" % (char, Decimal(float(occurance[char]) / count)))
	dump.close()

	print("unigram", language, corpus[language], count)


def bigram(language, file):
	text = open(file).read().lower()
	occurance = collections.Counter(find_ngrams(text, 2))
	count = len(find_ngrams(text, 2))

	filename = ("bigram_%s.txt" % language)
	dump = open(filename, "w")
	for char_1 in valid_chars:
		for char_2 in valid_chars:
			dump.write("'%s%s': %e,\n" % (char_2, char_1, Decimal(float(occurance[(char_2, char_1)]) / count)))
	dump.close()

	print("bigram", language, corpus[language], count)

corpus = {
	"danish": "dan_news_2008_30K-sentences.txt",
	"english": "eng_news_2016_30K-sentences.txt",
	"french": "fra_news_2010_30K-sentences.txt",
	"german": "deu_news_2015_30K-sentences.txt",
	"swedish": "swe_news_2007_30K-sentences.txt"
}

for language in corpus.keys():
	# unigram(language, corpus[language])
	bigram(language, corpus[language])
